File: dotnet/ArqLibrarianClassic/ArcLibrarianClassis.ATests/SpyUserOut.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using ArqLibrarianClassic;
using NUnit.Framework;

namespace ArcLibrarianClassis.ATests
{
    public class SpyUserOut : UserOut, InputAware
    {
        private readonly List<string> history = new List<string>();

        public SpyUserOut()
        {
            history.Add("");
        }

        public void Print(string text)
        {
            int lastElement = history.Count - 1;
            history[lastElement] = history[lastElement] + text;
        }

        public void PrintLine(string text)
        {
            Print(text);
            history.Add("");
        }

        internal void AssertCountainsAtLeastLines(int expectedCount)
        {
            if (history.Count < expectedCount)
            {
                Assert.Fail($"{Environment.NewLine}" +
                            $"UI contains less than {expectedCount} lines{Environment.NewLine}" +
                            $"{this.FormatHistory()}");
            }
        }

        internal void AssertContains(string expression)
        {
            if (history.Count == 0) 
            {
                Assert.Fail("No console entries at all. Especially: " + expression);
            }

            if (history.FindAll(x => Regex.IsMatch(x, expression)).Count == 0)
            {
                Assert.Fail($"{Environment.NewLine}" +
                            $"UI does not contain '{expression}'.{Environment.NewLine}" +
                            $"{this.FormatHistory()}");
            }
        }

        private string FormatHistory()
        {
            return $"History:{Environment.NewLine}" +
                   $"========={Environment.NewLine}" +
                   $"{string.Join($"{Environment.NewLine}", history)}";
        }

        public void OnTextLineEntered(string text)
        {
            PrintLine(text);
        }
    }
}

File: java/ArqLibraryClassic/ArqLibraryLayeredATests/src/bnsit/ms/arq/SpyUserOut.java
package bnsit.ms.arq;

import org.junit.Assert;

import java.util.ArrayList;
import java.util.List;

public class SpyUserOut implements InputAware, UserOut {

    private final List<String> history = new ArrayList<String>();

    public SpyUserOut() {
        history.add("");
    }

    @Override
    public void onTextLineEntered(String text) {
        printLine(text);
    }

    @Override
    public void print(String text) {
        int lastElement = history.size() - 1;
        history.set(lastElement, history.get(lastElement) + text);
    }

    @Override
    public void printLine(String text) {
        print(text);
        history.add("");
    }


    public void assertContains(String expression) {
        if (history.size() == 0)
        {
            Assert.fail("No console entries at all. Especially: " + expression);
        }

        if (history.stream().filter(h -> h.matches(".*" + expression + ".*")).count() == 0) {
            Assert.fail(String.format("%sUI does not contain '%s'.%s%s", newLine(), expression, newLine(), formatHistory()));

        }
    }

    public void assertContainsAtLeastLines(int expectedCount) {
        if (history.size() < expectedCount)
        {
            Assert.fail(String.format("%sUI contains less than %d lines%s%s", newLine(), expectedCount, newLine(), formatHistory()));
        }
    }

    private String formatHistory()
    {
        return String.format("History:%s=======%s%s", newLine(), newLine(), joinedHistory());
    }

    private String joinedHistory() {
        return String.join(newLine(), history);
    }

    private String newLine() {
        return System.getProperty("line.separator");
    }
}
